open_spiel/games/othello.cc

```diff
@@ -0,0 +1,344 @@
+  switch (state) {
+    case CellState::kEmpty:
+      return ".";
+    case CellState::kWhite:
+      return "o";
+    case CellState::kBlack:
+      return "x";
+    default:
+      SpielFatalError("Unknown state.");
+  }
+}
+
+bool OthelloState::ValidAction(Player player, int move) const {
+  return (board_[move] == CellState::kEmpty && CanCapture(player, move));
+}
+
+void OthelloState::DoApplyAction(Action move) {
+  if (move == passMove) {  // pass
+    current_player_ = 1 - current_player_;
+    num_moves_ += 1;
+    return;
+  }
+
+  if (!ValidAction(current_player_, move)) {
+    SpielFatalError(absl::StrCat("Invalid move ", move));
+  }
+
+  std::tuple<int, int> rowcol = XYFromCode(move);
+  CellState cell = PlayerToState(current_player_);
+
+  int row = std::get<0>(rowcol);
+  int col = std::get<1>(rowcol);
+  board_[row * kNumCols + col] = cell;
+  
+  for (int direction = Direction::kUp; direction < Direction::kLast; direction++) {
+    int steps = 0;
+    if ((steps = CountSteps(current_player_, row, col, static_cast<Direction>(direction))) > 0) {
+      Capture(current_player_, row, col, static_cast<Direction>(direction), steps);
+    }
+  }
+
+  if (NoValidActions()) {
+    int count_zero = DiskCount(Player(0));
+    int count_one = DiskCount(Player(1));
+
+    if (count_zero > count_one) {
+      outcome_ = Player(0);
+    } else if (count_zero < count_one) {
+      outcome_ = Player(1);
+    }
+  }
+
+  current_player_ = 1 - current_player_;
+  num_moves_ += 1;
+}
+
+std::vector<Action> OthelloState::LegalRegularActions(Player p) const {  // list 
+  std::vector<Action> moves;
+  for (int cell = 0; cell < kNumCells; ++cell) {
+    if (ValidAction(p, cell)) {
+      moves.push_back(cell);
+    }
+  }
+
+  return moves;
+}
+
+std::vector<Action> OthelloState::LegalActions() const {
+  if (IsTerminal()) return {};
+  
+  // can move in an empty cell that captures
+  std::vector<Action> moves = LegalRegularActions(current_player_);
+  if (moves.empty()) {
+    moves.push_back(passMove);  // pass
+  }
+
+  return moves;
+}
+
+std::string OthelloState::ActionToString(Player player,
+                                           Action action_id) const {
+
+  if (action_id == passMove) {
+    return absl::StrCat(StateToString(PlayerToState(player)), "(pass)");
+  } else {
+    return absl::StrCat(StateToString(PlayerToState(player)), "(",
+                      action_id / kNumCols, ",", action_id % kNumCols, ")");
+  }
+}
+
+bool OthelloState::IsFull() const { 
+  for (int i = 0; i < kNumCells; i++) {
+    if (board_[i] == CellState::kEmpty) return false;
+  }
+
+  return true;
+}
+
+OthelloState::OthelloState(std::shared_ptr<const Game> game) : State(game) {
+  std::fill(begin(board_), end(board_), CellState::kEmpty);
+  board_[27] = CellState::kWhite;
+  board_[28] = CellState::kBlack;
+  board_[35] = CellState::kBlack;
+  board_[36] = CellState::kWhite;
+}
+
+std::string OthelloState::ToString() const {
+  std::string str;
```

Comment:
It could be useful to put the coordinate system on here too, especially if you make the suggested change of action string.

Reply:
Added.